lib/types.nix

```diff
@@ -367,22 +367,47 @@ rec {
         name = "coercedTo";
         description = "${finalType.description} or ${coercedType.description}";
         check = x: finalType.check x || coercedType.check x;
-        merge = loc: defs:
+        merge =
           let
             coerceVal = val:
               if finalType.check val then val
               else let
                 coerced = coerceFunc val;
               in assert finalType.check coerced; coerced;
 
-          in finalType.merge loc (map (def: def // { value = coerceVal def.value; }) defs);
+          in loc: defs: finalType.merge loc (map (def: def // { value = coerceVal def.value; }) defs);
         getSubOptions = finalType.getSubOptions;
         getSubModules = finalType.getSubModules;
         substSubModules = m: coercedTo coercedType coerceFunc (finalType.substSubModules m);
         typeMerge = t1: t2: null;
         functor = (defaultFunctor name) // { wrapped = finalType; };
       };
 
+    # Function from `from` to `to`.
+    function = from: to:
+      assert from.getSubModules == null;
+      mkOptionType rec {
+        name = "function";
```

Comment:
Ah, thanks for the reference!

I'm unsure how to do it the other way (avoid using functions). Without them every added package would need to be effectively doubled, e.g. we'd need to have two lists of identical packages but for different architecture specified by users (we in fact have this now as `hardware.opengl.extraPackages{,32}` and it looks quite ugly). Also I think a function is just a very natural way to express this idea (abstract away from arch, and focus on "what packages are needed").

Any ideas much appreciated!